### PR TITLE
migrate to separate uTP crate

### PR DESCRIPTION
**Pull request by jacobkaufmann**
### What was wrong?

`trin` maintains an internal uTP implementation based on the [`utp`](https://crates.io/crates/utp) crate. This approach was taken to support uTP in a timely manner, but now we have the opportunity to replace it with a more general-purpose library that is not so interwoven into other system components.

### How was it fixed?

Changes:

* remove the `trin_core::utp` module
* add a dependency on `utp-rs`
* use `UtpSocket` in place of `UtpListener`
* implement an abstract UDP socket over Discovery v5
* update `utp-testing` to use the transport from `utp-rs`

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

**Comment by jacobkaufmann** [comment]
the version of `utp-rs` here is `0.1.0-alpha` for now. I would like to collect some feedback before moving `utp-rs` beyond a pre-release version. once I have confidence that the API is stable, I will make a `0.1.0` release and update the PR to depend on that version.

File: .circleci/config.yml
Comment (jacobkaufmann) [inline .circleci/config.yml:125]: note: this change was made because the 2nd packet corresponds to the node's response to the discovery v5 `WHOAREYOU` challenge

File: trin-core/src/portalnet/discovery.rs
Comment (ogenev) [inline trin-core/src/portalnet/discovery.rs:309]: Is this `tokio::sync::Mutex` instead of `RwLock`?

**Review by ogenev — COMMENTED** [review]
Looks solid to me, I think it is a great improvement over what we had. :+1: 

Did we test OFFER/ACCEPT and FINDCONTENT?CONTENT compatibility with Fluffy?

Sadly, we don't have automatic tests for this and before approval, I think we should test this manually.

File: trin-core/src/portalnet/discovery.rs
Comment (jacobkaufmann) [reply trin-core/src/portalnet/discovery.rs:309]: good catch! it was a `tokio::sync::RwLock` initially. I will update the comment.

File: .circleci/config.yml
Comment (jacobkaufmann) [reply .circleci/config.yml:125]: at this time I'm not sure, but I can look into it further.

**Comment by jacobkaufmann** [comment]
I did some manual testing by running a bridge peered with a (uTP-modified) `trin` node, but I should do some more isolated testing with a local `fluffy` node. Once confirmed successful, I will leave a note here.

File: trin-core/src/portalnet/overlay_service.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/overlay_service.rs:1106]: `if let Err(err) = stream.read_to_eof(&mut data).await {..}`?

File: trin-core/src/portalnet/discovery.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/discovery.rs:325]: Can you leave a comment on why this additional enr wrapped type is necessary

File: .circleci/config.yml
Comment (carver) [reply .circleci/config.yml:125]: Any luck tracking this down?

File: ethportal-peertest/src/scenarios/offer_accept.rs
Comment (carver) [inline ethportal-peertest/src/scenarios/offer_accept.rs:103]: A code comment explaining the intuition of why not to drop it to 0.5s would be nice.

**Review by carver — APPROVED** [review]
Hooray for a bunch of deleted code! 👍🏻 for the direction

> I should do some more isolated testing with a local fluffy node. Once confirmed successful, I will leave a note here.

I agree this is worth doing.

File: ethportal-peertest/src/scenarios/offer_accept.rs
Comment (jacobkaufmann) [reply ethportal-peertest/src/scenarios/offer_accept.rs]: I was able to move the retry duration back to 0.5 seconds with no failure.

File: .circleci/config.yml
Comment (jacobkaufmann) [reply .circleci/config.yml:125]: unfortunately I have not been able to track this down. I'm not sure why things were working before. that said, I do not think this should block merge because we move the drop into the uTP data transfer, which is the aspect we are interested in testing.

**Comment by jacobkaufmann** [comment]
I ran into some discv5-level issues testing a local `trin` and `fluffy` pair, but I have been able to peer a local `trin` node with the public portal network and successfully transmit data via uTP. I believe that we are in a place to move forward and diagnose/address issues as they crop up. there are further enhancements to make to automated uTP testing, but that is outside the present scope.

**Pull request merged**
